### PR TITLE
Gemma 7B x LoRA fix

### LINKED ISSUE #1124: Gemma 7B Lora Bug
There seems to be a bug when using Gemma-7B with certain LoRA settings:

```
litgpt finetune lora --config litgpt/config_hub/finetune/gemma-7b/lora.yaml


{'checkpoint_dir': PosixPath('checkpoints/google/gemma-7b'),
 'data': Alpaca2k(mask_prompt=False, val_split_fraction=0.05, prompt_style=<litgpt.prompts.Alpaca object at 0x7f7ffa33f310>, ignore_index=-100, seed=42, num_workers=4, download_dir=PosixPath('data/alpaca2k')),
 'devices': 1,
 'eval': EvalArgs(interval=100, max_new_tokens=100, max_iters=100),
 'logger_name': 'csv',
 'lora_alpha': 16,
 'lora_dropout': 0.05,
 'lora_head': False,
 'lora_key': False,
 'lora_mlp': False,
 'lora_projection': False,
 'lora_query': True,
 'lora_r': 32,
 'lora_value': True,
 'out_dir': PosixPath('out/finetune/lora-gemma-7b'),
 'precision': 'bf16-true',
 'quantize': None,
 'seed': 1337,
 'train': TrainArgs(save_interval=200, log_interval=1, global_batch_size=8, micro_batch_size=2, lr_warmup_steps=10, epochs=4, max_tokens=None, max_steps=None, max_seq_length=512, tie_embeddings=None, learning_rate=0.0002, weight_decay=0.0, beta1=0.9, beta2=0.95, max_norm=None, min_lr=6e-05)}
Seed set to 1337
Number of trainable parameters: 11,010,048
Number of non trainable parameters: 9,324,112,896
The longest sequence length in the train data is 512, the model's maximum sequence length is 512 and context length is 4096
Validating ...
Traceback (most recent call last):
  File "/home/zeus/miniconda3/envs/cloudspace/bin/litgpt", line 8, in <module>
    sys.exit(main())
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/litgpt/__main__.py", line 131, in main
    fn(**kwargs)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/litgpt/finetune/lora.py", line 137, in setup
    fabric.launch(main, devices, seed, config, data, checkpoint_dir, out_dir, train, eval)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/lightning/fabric/fabric.py", line 866, in launch
    return self._wrap_and_launch(function, self, *args, **kwargs)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/lightning/fabric/fabric.py", line 952, in _wrap_and_launch
    return to_run(*args, **kwargs)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/lightning/fabric/fabric.py", line 957, in _wrap_with_setup
    return to_run(*args, **kwargs)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/litgpt/finetune/lora.py", line 181, in main
    fit(fabric, model, optimizer, scheduler, train_dataloader, val_dataloader, devices, checkpoint_dir, out_dir, train, eval, data)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/litgpt/finetune/lora.py", line 219, in fit
    validate(fabric, model, val_dataloader, tokenizer, dataclasses.replace(eval, max_iters=2), data)  # sanity check
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/litgpt/finetune/lora.py", line 317, in validate
    logits = model(input_ids)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/lightning/fabric/wrappers.py", line 143, in forward
    output = self._forward_module(*args, **kwargs)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/litgpt/lora.py", line 545, in forward
    x = block(x, cos, sin, mask, input_pos)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/litgpt/model.py", line 158, in forward
    h = self.attn(n_1, cos, sin, mask, input_pos)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/litgpt/model.py", line 197, in forward
    qkv = self.attn(x)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/litgpt/lora.py", line 432, in forward
    lora = self.zero_pad(after_B) * self.scaling  # (64, 64, 256) after zero_pad (64, 64, 384)
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.9/site-packages/litgpt/lora.py", line 339, in zero_pad
    result = result.index_copy(
IndexError: index_copy_(): Number of indices (8192) should be equal to source.size(dim) (6144)
```

The config file for reference:

```
# The path to the base model's checkpoint directory to load for finetuning. (type: <class 'Path'>, default: checkpoints/stabilityai/stablelm-base-alpha-3b)
checkpoint_dir: checkpoints/google/gemma-7b

# Directory in which to save checkpoints and logs. (type: <class 'Path'>, default: out/lora)
out_dir: out/finetune/lora-gemma-7b

# The precision to use for finetuning. Possible choices: "bf16-true", "bf16-mixed", "32-true". (type: Optional[str], default: null)
precision: bf16-true

# If set, quantize the model with this algorithm. See ``tutorials/quantize.md`` for more information. (type: Optional[Literal['nf4', 'nf4-dq', 'fp4', 'fp4-dq', 'int8-training']], default: null)
quantize:

# How many devices/GPUs to use. (type: Union[int, str], default: 1)
devices: 1

# The LoRA rank. (type: int, default: 8)
lora_r: 32

# The LoRA alpha. (type: int, default: 16)
lora_alpha: 16

# The LoRA dropout value. (type: float, default: 0.05)
lora_dropout: 0.05

# Whether to apply LoRA to the query weights in attention. (type: bool, default: True)
lora_query: true

# Whether to apply LoRA to the key weights in attention. (type: bool, default: False)
lora_key: false

# Whether to apply LoRA to the value weights in attention. (type: bool, default: True)
lora_value: true

# Whether to apply LoRA to the output projection in the attention block. (type: bool, default: False)
lora_projection: false

# Whether to apply LoRA to the weights of the MLP in the attention block. (type: bool, default: False)
lora_mlp: false

# Whether to apply LoRA to output head in GPT. (type: bool, default: False)
lora_head: false

# Data-related arguments. If not provided, the default is ``litgpt.data.Alpaca``.
data:
  class_path: litgpt.data.Alpaca2k
  init_args:
    mask_prompt: false
    prompt_style: alpaca
    ignore_index: -100
    seed: 42
    num_workers: 4

# Training-related arguments. See ``litgpt.args.TrainArgs`` for details
train:

  # Number of optimizer steps between saving checkpoints (type: Optional[int], default: 1000)
  save_interval: 200

  # Number of iterations between logging calls (type: int, default: 1)
  log_interval: 1

  # Number of samples between optimizer steps across data-parallel ranks (type: int, default: 128)
  global_batch_size: 8

  # Number of samples per data-parallel rank (type: int, default: 4)
  micro_batch_size: 2

  # Number of iterations with learning rate warmup active (type: int, default: 100)
  lr_warmup_steps: 10

  # Number of epochs to train on (type: Optional[int], default: 5)
  epochs: 4

  # Total number of tokens to train on (type: Optional[int], default: null)
  max_tokens:

  # Limits the number of optimizer steps to run. (type: Optional[int], default: null)
  max_steps:

  # Limits the length of samples. Off by default (type: Optional[int], default: null)
  max_seq_length: 512

  # Whether to tie the embedding weights with the language modeling head weights. (type: Optional[bool], default: null)
  tie_embeddings:

  #   (type: float, default: 0.0003)
  learning_rate: 0.0002

  #   (type: float, default: 0.02)
  weight_decay: 0.0

  #   (type: float, default: 0.9)
  beta1: 0.9

  #   (type: float, default: 0.95)
  beta2: 0.95

  #   (type: Optional[float], default: null)
  max_norm:

  #   (type: float, default: 6e-05)
  min_lr: 6.0e-05

# Evaluation-related arguments. See ``litgpt.args.EvalArgs`` for details
eval:

  # Number of optimizer steps between evaluation calls (type: int, default: 100)
  interval: 100

  # Number of tokens to generate (type: Optional[int], default: 100)
  max_new_tokens: 100

  # Number of iterations (type: int, default: 100)
  max_iters: 100

# The name of the logger to send metrics to. (type: Literal['wandb', 'tensorboard', 'csv'], default: csv)
logger_name: csv

# The random seed to use for reproducibility. (type: int, default: 1337)
seed: 1337
```

### PR DESCRIPTION
**Pull request by Andrei-Aksionov**
Hi there 👋 

This PR fixes an issue with Gemma 7B variant and LoRA. The problem is that for the 7B variant `n_emd` != `head_size` * `n_head` and thus that caused an error when calculating the shape for `lora_B`.

Fixes #1124

### CODE REVIEW

File: litgpt/lora.py [hunk 1]
```diff
@@ -235,12 +237,13 @@ def __init__(
         if r > 0 and any(enable_lora):
             self.lora_A = nn.Parameter(torch.zeros((r * sum(enable_lora), in_features)))  # (4, 128)
             enable_q, enable_k, enable_v = enable_lora
-            self.kv_embd_size = self.linear.in_features // (n_head // n_query_groups)
             # qkv_shapes will be used to split a tensor with weights correctly
             qkv_shapes = (
-                self.linear.in_features * enable_q,
-                self.kv_embd_size * enable_k,
-                self.kv_embd_size * enable_v,
+                # if `head_size` is explicitly specified in the config, `n_emd` (or `in_features`)
```
Comment (carmocca) [inline litgpt/lora.py:242]:
```suggestion
                # if `head_size` is explicitly specified in the config, `n_embd` (or `in_features`)
```

File: tests/test_lora.py [hunk 2]
```diff
@@ -524,6 +528,9 @@ def test_against_hf_mixtral():
         n_query_groups=2,
         intermediate_size=86,
         n_expert=4,
+        lora_r=1,
+        lora_key=True,
+        lora_value=True,
```
Comment (carmocca) [inline tests/test_lora.py:533]: What's the motivation for adding these? Shouldn't the end to end results fail now?

Comment (Andrei-Aksionov) [reply tests/test_lora.py:533]: By default, `lora_r` is 0 which basically makes it a regular GPT, not LoRA variant, since we skip LoRA path.

Comment (carmocca) [reply tests/test_lora.py:533]: Right but the test compares the numerical output to the regular model from HuggingFace

Comment (Andrei-Aksionov) [reply tests/test_lora.py:533]: Yes :)
The `lora_B` weights are initialized with zeros, so `lora_A` @ `lora_B` produces a matrix filled with zeros.
Pretrained + [0, 0, ..., 0] = Pretrained, thus it allows us to compare to the regular model.
Here we check this assumption plus verify that the code works properly.

Comment (carmocca) [reply tests/test_lora.py:533]: Ahh I forgot that the LoRA __init__ calls `reset_parameters`. Thank you LoRA expert

**Pull request merged**